packages/graphql/src/schema/validation/custom-rules/features/valid-global-id.ts

```diff
@@ -0,0 +1,277 @@
+    interfaceToImplementingTypes: Map<string, string[]>
+): string[] | undefined {
+    if (mainType.kind === Kind.INTERFACE_TYPE_DEFINITION) {
+        return interfaceToImplementingTypes.get(mainType.name.value);
+    }
+    if (mainType.kind === Kind.OBJECT_TYPE_DEFINITION) {
+        return mainType.interfaces?.map((i) => i.name.value);
+    }
+}
+
+function hasGlobalIDField(
+    mainType: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
+    typeNameToGlobalId: Map<string, boolean>,
+    interfaceToImplementingTypes: Map<string, string[]>
+): boolean {
+    if (typeNameToGlobalId.get(mainType.name.value)) {
+        return true;
+    }
+    return !!getInheritedTypeNames(mainType, interfaceToImplementingTypes)?.find(
+        (typeName) => typeNameToGlobalId.get(typeName) === true
+    );
+}
+
+function assertGlobalIDDoesNotClash(aliasedFieldsFromInheritedTypes: (Set<string> | undefined)[]) {
+    let shouldDeferCheck: boolean | undefined = false;
+    let hasAlias: boolean | undefined = false;
+    for (const aliasedFields of aliasedFieldsFromInheritedTypes) {
+        if (!aliasedFields) {
+            shouldDeferCheck = true;
+        } else {
+            if (aliasedFields.has("id")) {
+                hasAlias = true;
+                return;
+            }
+        }
+    }
+    if (hasAlias === false && shouldDeferCheck === false) {
+        throw new DocumentValidationError(
+            'Invalid global id field: Types decorated with an `@id` directive with the global argument set to `true` cannot have a field named "id". Either remove it, or if you need access to this property, consider using the "@alias" directive to access it via another field.',
+            ["id"]
+        );
+    }
+}
+
+function assertValidGlobalID({
+    directiveNode,
+    typeDef,
+    typeNameToGlobalId,
+    interfaceToImplementingTypes,
+}: {
+    directiveNode: DirectiveNode;
+    typeDef: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode;
+    typeNameToGlobalId: Map<string, boolean>;
+    interfaceToImplementingTypes: Map<string, string[]>;
+}) {
+    if (hasGlobalIDField(typeDef, typeNameToGlobalId, interfaceToImplementingTypes)) {
+        throw new DocumentValidationError(
+            "Invalid directive usage: Only one field may be decorated with an '@id' directive with the global argument set to `true`.",
+            ["@id", "global"]
+        );
+    } else {
+        typeNameToGlobalId.set(typeDef.name.value, true);
+    }
+
+    const isNotUnique = directiveNode.arguments?.find(
+        (a) => a.name.value === "unique" && a.value.kind === Kind.BOOLEAN && a.value.value === false
+    );
+    if (isNotUnique) {
+        throw new DocumentValidationError(
+            `Invalid global id field: Fields decorated with the "@id" directive must be unique in the database. Please remove it, or consider making the field unique.`,
+            ["@id", "unique"]
+        );
+    }
```

Comment:
Trying to make sense of this - what is this rule?

Reply:
well they are by default :D `@id` has the `unique` argument as `true` by default

Reply:
But the "Invalid global id field" is confusing - it makes it sound like all fields decorated with `@id` are global ID fields, which isn't the case. I would go with something like `Invalid global ID field - unique argument must be set to true if global argument is set to true` - or something or other.